### PR TITLE
Add test case for goc diff

### PR DESCRIPTION
**Pull request by tongjingran**

### CODE REVIEW

File: pkg/prow/job.go [hunk 1]
```diff
@@ -231,3 +209,31 @@ func (j *Job) CreateChangedCovHtml() error {
 	}
 	return err
 }
+
+func getFilesAndCovList(fullDiff bool, prComment github.PrComment, localP, baseP cover.CoverageList) (changedFiles []string, deltaCovList cover.DeltaCovList, err error) {
+	if !fullDiff {
+		// get github pull request changed files' name
+		var ghChangedFiles, err = prComment.GetPrChangedFiles()
+		if err != nil {
+			logrus.Errorf("Get pull request changed file failed.")
```
Comment (CarlJi) [inline pkg/prow/job.go:218]: format `err` with this message and return to upstream seems better

File: pkg/prow/job.go [hunk 2]
```diff
@@ -231,3 +209,31 @@ func (j *Job) CreateChangedCovHtml() error {
...
+			return nil, nil, nil
+		}
+		changedFiles = trimGhFileToProfile(ghChangedFiles)
+
+		// calculate diff cov between local and remote profile
+		deltaCovList = cover.GetChFileDeltaCov(localP, baseP, changedFiles)
+		return changedFiles, deltaCovList, nil
+	}
+	deltaCovList = cover.GetDeltaCov(localP, baseP)
+	logrus.Infof("get delta file name is:")
```
Comment (CarlJi) [inline pkg/prow/job.go:232]: suggest to remove this statement

File: pkg/prow/job.go [hunk 3]
```diff
@@ -231,3 +209,31 @@ func (j *Job) CreateChangedCovHtml() error {
...
+	for _, d := range deltaCovList {
+		logrus.Infof("%s", d.FileName)
```
Comment (CarlJi) [inline pkg/prow/job.go:234]: remove too

File: pkg/prow/job.go [hunk 4]
```diff
@@ -231,3 +209,31 @@ func (j *Job) CreateChangedCovHtml() error {
...
+		changedFiles = append(changedFiles, d.FileName)
+	}
+
```
Comment (CarlJi) [inline pkg/prow/job.go:237]: if necessary, it can print a information here

File: pkg/prow/job_test.go [hunk 5]
```diff
@@ -17,16 +17,28 @@
 package prow
 
 import (
+	"errors"
```
Comment (CarlJi) [inline pkg/prow/job_test.go:20]: format the imports list based on https://gist.github.com/adamveld12/c0d9f0d5f0e1fba1e551#imports

File: pkg/prow/job_test.go [hunk 6]
```diff
@@ -130,12 +136,176 @@ func TestRunPresubmitFulldiff(t *testing.T) {
...
 	defer os.Remove(path.Join(os.Getenv("ARTIFACTS"), j.HtmlProfile()))
 
-	j.RunPresubmit()
+	err = j.RunPresubmit()
+	assert.NoError(t, err)
+}
+
+func TestRunPresubmitError(t *testing.T) {
+	items := []struct {
+		prepare bool // 是否需要准备本地cov
```
Comment (CarlJi) [inline pkg/prow/job_test.go:152]: translate to english

File: pkg/qiniu/client.go [hunk 7]
```diff
@@ -43,23 +43,32 @@ type Config struct {
 	Domain string `json:"domain"`
 }
 
-// Client for the operation with qiniu cloud
-type Client struct {
+type Client interface {
```
Comment (CarlJi) [inline pkg/qiniu/client.go:46]: suggest to add comments for this exported interface

File: pkg/qiniu/object.go [hunk 8]
```diff
@@ -29,8 +29,14 @@ import (
 	"github.com/sirupsen/logrus"
 )
 
-// ObjectHandle provides operations on an object in a qiniu cloud bucket
-type ObjectHandle struct {
+type ObjectHandle interface {
```
Comment (CarlJi) [inline pkg/qiniu/object.go:32]: suggest to add comments for exported interface

File: pkg/qiniu/qnPresubmit.go [hunk 9]
```diff
@@ -107,20 +107,26 @@ func FindBaseProfileFromQiniu(qc *Client, prowJobName, covProfileName string) ([
 	return qc.ReadObject(profilePath)
 }
 
-// Artifacts prepresents the rule to store test artifacts in prow
-type Artifacts struct {
+type Artifacts interface {
```
Comment (CarlJi) [inline pkg/qiniu/qnPresubmit.go:110]: suggest to add comments for exported interface

File: pkg/qiniu/qnPresubmit.go [hunk 10]
```diff
@@ -132,3 +138,7 @@ func (a *Artifacts) CreateChangedProfile() *os.File {
 
 	return p
 }
+
+func (a *ProfileArtifacts) GetChangedProfileName() string {
```
Comment (CarlJi) [inline pkg/qiniu/qnPresubmit.go:142]: suggest to add comments for exported method

File: pkg/qiniu/qnPresubmit_test.go [hunk 11]
```diff
@@ -20,6 +20,7 @@ import (
 	"testing"
 
 	"github.com/stretchr/testify/assert"
+	"os"
```
Comment (CarlJi) [inline pkg/qiniu/qnPresubmit_test.go:23]: format import list based on https://gist.github.com/adamveld12/c0d9f0d5f0e1fba1e551#imports

File: pkg/qiniu/qnPresubmit_test.go [hunk 12]
```diff
@@ -39,3 +40,29 @@ func TestFindBaseProfileFromQiniu(t *testing.T) {
...
+func TestArtifacts_ProfilePath(t *testing.T) {
+	p := &ProfileArtifacts{
+		Directory:   "directory/",
+		ProfileName: "profile",
+	}
+	profilePath := p.ProfilePath()
+	assert.Equal(t, profilePath, "directory/profile")
+}
+
+func TestProfileArtifacts_CreateChangedProfile(t *testing.T) {
```
Comment (CarlJi) [inline pkg/qiniu/qnPresubmit_test.go:53]: seems no assert action

File: pkg/prow/job.go (hunk 1 above)
Comment (tongjingran) [reply pkg/prow/job.go]: fixed~

File: pkg/prow/job.go (hunk 2 above)
Comment (tongjingran) [reply pkg/prow/job.go]: removed~

File: pkg/prow/job.go (hunk 3 above)
Comment (tongjingran) [reply pkg/prow/job.go]: removed~

File: pkg/prow/job.go (hunk 4 above)
Comment (tongjingran) [reply pkg/prow/job.go]: add info before return

File: pkg/qiniu/qnPresubmit_test.go (hunk 12 above)
Comment (tongjingran) [reply pkg/qiniu/qnPresubmit_test.go:53]: add an assert action

File: pkg/qiniu/qnPresubmit_test.go (hunk 11 above)
Comment (tongjingran) [reply pkg/qiniu/qnPresubmit_test.go]: fixed~

File: pkg/qiniu/qnPresubmit.go (hunk 10 above)
Comment (tongjingran) [reply pkg/qiniu/qnPresubmit.go:143]: added~

File: pkg/qiniu/qnPresubmit.go (hunk 9 above)
Comment (tongjingran) [reply pkg/qiniu/qnPresubmit.go:111]: added~

File: pkg/qiniu/object.go (hunk 8 above)
Comment (tongjingran) [reply pkg/qiniu/object.go:33]: added~

File: pkg/qiniu/client.go (hunk 7 above)
Comment (tongjingran) [reply pkg/qiniu/client.go:47]: added~

File: pkg/prow/job_test.go (hunk 5 above)
Comment (tongjingran) [reply pkg/prow/job_test.go:20]: fixed~

File: pkg/qiniu/object.go [hunk 13]
```diff
@@ -29,8 +29,15 @@ import (
 	"github.com/sirupsen/logrus"
 )
 
-// ObjectHandle provides operations on an object in a qiniu cloud bucket
-type ObjectHandle struct {
+// ObjectHandle is the interface contains the operations on an object in a qiniu cloud bucket
+type ObjectHandle interface {
+	Attrs(ctx context.Context) (storage.FileInfo, error)
```
Comment (lyyyuna) [inline pkg/qiniu/object.go:34]: 这个interface能否剥离对 storage.FileInfo 的依赖

File: pkg/qiniu/qnPresubmit.go (hunk 9 above)
Comment (CarlJi) [reply pkg/qiniu/qnPresubmit.go:111]: `therule` => `the rule`

File: pkg/qiniu/object.go (hunk 13 above)
Comment (tongjingran) [reply pkg/qiniu/object.go]: remove this method

File: pkg/prow/job.go [hunk 14]
```diff
@@ -67,77 +67,55 @@ func (j *Job) Fetch(BuildID, name string) []byte {
...
-		}
-		changedFiles = trimGhFileToProfile(ghChangedFiles)
-	}
-
-	// step2: get local profile cov
+	// step1: get local profile cov
 	localP, err := cover.ReadFileToCoverList(j.LocalProfilePath)
 	if err != nil {
-		logrus.WithError(err).Fatalf("failed to get remote cover profile")
+		return fmt.Errorf("failed to get remote cover profile: %s", err.Error())
```
Comment (lyyyuna) [inline pkg/prow/job.go:73]: Wrapping errors with %w

File: pkg/prow/job.go [hunk 15]
```diff
@@ -231,3 +209,29 @@ func (j *Job) CreateChangedCovHtml() error {
 	}
 	return err
 }
+
+func getFilesAndCovList(fullDiff bool, prComment github.PrComment, localP, baseP cover.CoverageList) (changedFiles []string, deltaCovList cover.DeltaCovList, err error) {
+	if !fullDiff {
+		// get github pull request changed files' name
+		var ghChangedFiles, err = prComment.GetPrChangedFiles()
+		if err != nil {
+			return nil, nil, fmt.Errorf("Get pull request changed file failed: %s", err.Error())
```
Comment (lyyyuna) [inline pkg/prow/job.go:218]: Wrapping errors with %w

**Pull request merged**